Xamarin.Forms.Core/IndicatorsView.cs

```diff
@@ -0,0 +1,285 @@
+				Source = carouselView
+			});
+		}
+
+		protected virtual void ApplySelectedStyle(View view, int index)
+			=> view.BackgroundColor = SelectedIndicatorColor;
+
+		protected virtual void ApplyUnselectedStyle(View view, int index)
+			=> view.BackgroundColor = IndicatorsColor;
+
+		private void AddExtraIndicatorsItems()
+		{
+			var oldCount = Children.Count;
+			for (var i = 0; i < Count - oldCount && i < MaxVisibleCount - oldCount; ++i)
+			{
+				var indicator = IndicatorTemplate?.CreateContent() as View ?? new Frame
+				{
+					Padding = 0,
+					HasShadow = false,
+					BorderColor = Color.Transparent,
+					VerticalOptions = LayoutOptions.Center,
+					HorizontalOptions = LayoutOptions.Center,
+					WidthRequest = IndicatorsSize,
+					HeightRequest = IndicatorsSize,
+					CornerRadius = IndicatorShape == IndicatorShape.Circle
+						? (float)IndicatorsSize / 2
+						: 0
+				};
+				var itemTapGesture = new TapGestureRecognizer();
+				itemTapGesture.Tapped += (tapSender, tapArgs) => Position = IndexOf(tapSender as View);
+				indicator.GestureRecognizers.Add(itemTapGesture);
+				Children.Add(indicator);
+			}
+		}
+
+		private void RemoveRedundantIndicatorsItems()
+		{
+			foreach (var item in Children.Where((v, i) => i >= Count).ToArray())
+			{
+				Children.Remove(item);
+			}
+		}
+
+		private int IndexOf(View view) => Children.IndexOf(view);
+
+		private void ApplyStyle(View view)
+		{
+			try
+			{
+				view.BatchBegin();
+				var index = IndexOf(view);
+				if (index == Position)
+				{
+					ApplySelectedStyle(view, index);
+					return;
+				}
+				ApplyUnselectedStyle(view, index);
+			}
+			finally
+			{
+				view.BatchCommit();
+			}
+		}
+
+		private void ResetIndicatorsStylesNonBatch()
+		{
+			foreach (var child in Children)
+			{
+				ApplyStyle(child);
+			}
+		}
+
+		private void ResetIndicatorsStyles()
+		{
+			try
+			{
+				BatchBegin();
+				ResetIndicatorsStylesNonBatch();
+			}
+			finally
+			{
+				BatchCommit();
+			}
+		}
+
+		private void ResetIndicatorsCount(int oldValue, int newValue)
+		{
+			try
+			{
+				BatchBegin();
+				if (oldValue < 0)
+				{
+					oldValue = 0;
+				}
+
+				if (oldValue > newValue)
+				{
+					RemoveRedundantIndicatorsItems();
+					return;
+				}
+
+				AddExtraIndicatorsItems();
+			}
+			finally
+			{
+				ResetIndicatorsStylesNonBatch();
+				BatchCommit();
+			}
+		}
+
+		private void ResetItemsSource(IEnumerable oldCollection)
+		{
+			if (oldCollection is INotifyCollectionChanged oldObservableCollection)
+			{
+				oldObservableCollection.CollectionChanged -= OnCollectionChanged;
+			}
+
+			if (ItemsSource is INotifyCollectionChanged observableCollection)
+			{
+				observableCollection.CollectionChanged += OnCollectionChanged;
+			}
+
+			OnCollectionChanged(ItemsSource, new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Reset));
+		}
+
+		private void OnCollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
```

Comment:
we can do better than that, this doesn't handle elements moved in a collection

Reply:
Hmmmm, why?

if ItemsSource is INotifyCollectionChanged it means that CollectionChanged event will be fired with **Move** action, in case of any element is moved

Am i wrong?